File: LWidget/src/LLineedit.cpp
#include "LLineedit.h"

#include <QKeyEvent>
#include <QTimer>

using namespace ljz;

LFileLineEdit::Info::Info(QFileDialog::FileMode mode, const QString& path, const QString& title,
	const QStringList& filters, const QIcon& icon)
	:mode(mode), path(path), title(title), filters(filters), icon(icon)
{
}

LFileLineEdit::LFileLineEdit(QWidget* parent)
{
	this->setReadOnly(true);
	this->_clearAction->setIcon(QIcon(":/res/res/clear.png"));
	this->_clearAction->setToolTip("清空");
	this->addAction(this->_clearAction, QLineEdit::TrailingPosition);
	connect(this->_clearAction, &QAction::triggered, [=]()
		{
			this->clear();
			this->setToolTip("");
			emit this->fileCleared();
		});
	reAction();
}

LFileLineEdit::LFileLineEdit(const QString& text, QWidget* parent)
	:QLineEdit(text, parent)
{
}

QAction* LFileLineEdit::action() const
{
	return this->_action;
}

QAction* LFileLineEdit::clearAction() const
{
	return this->_clearAction;
}

LFileLineEdit::Info LFileLineEdit::info() const
{
	return this->_info;
}

void LFileLineEdit::setInfo(const Info& info)
{
	this->_info = info;
	this->reAction();
}

QStringList LFileLineEdit::selectedFiles() const
{
	return this->text().split(";");
}

QString LFileLineEdit::selectedFiles(const QString& sep) const
{
	return this->selectedFiles().join(sep);
}

void LFileLineEdit::setClearActionVisible(bool visible) const
{
	this->_clearAction->setVisible(visible);
}

bool LFileLineEdit::clearActionVisible() const
{
	return this->_clearAction->isVisible();
}

void LFileLineEdit::showFileDialog()
{
	QFileDialog* fileDialog = new QFileDialog(nullptr, this->_info.title, this->_info.path);
	fileDialog->setNameFilters(_info.filters);
	fileDialog->setFileMode(this->_info.mode);
	if (fileDialog->exec() == QDialog::Accepted)
	{
		QStringList list = fileDialog->selectedFiles();
		QString filter = fileDialog->selectedNameFilter();
		if (!list.isEmpty())
		{
			this->setText(list.join(";"));
			this->setToolTip(list.join("\n"));
			emit this->fileSelected(list, filter);
		}
	}
}

void LFileLineEdit::reAction()
{
	delete this->_action;
	this->_action = new QAction();
	this->_action->setIcon(_info.icon);
	this->setPlaceholderText(_info.title);
	this->_action->setToolTip(_info.title);
	this->addAction(this->_action, QLineEdit::TrailingPosition);
	connect(this->_action, &QAction::triggered, this, &LFileLineEdit::showFileDialog);
}

LFocusSelectLineEdit::LFocusSelectLineEdit(QWidget* parent)
	:QLineEdit(parent)
{
}

LFocusSelectLineEdit::LFocusSelectLineEdit(const QString& text, QWidget* parent)
	:QLineEdit(text, parent)
{
}

void LFocusSelectLineEdit::focusInEvent(QFocusEvent* event)
{
	QLineEdit::focusInEvent(event);
	QTimer::singleShot(100, this, SLOT(selectAll()));
}

LHostAddressValidator::LHostAddressValidator(QObject* parent)
	:QValidator(parent)
{
}

LHostAddressValidator::~LHostAddressValidator() = default;

QValidator::State LHostAddressValidator::validate(QString& input, int& pos) const
{
	QRegularExpression regularExpression("^((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])?\\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])?$");

	QRegularExpressionMatch match = regularExpression.match(input);
	if (match.hasMatch())
	{
		return QValidator::State::Acceptable;
	}
	return QValidator::State::Invalid;
}

LHostAddressEdit::LHostAddressEdit(const QHostAddress& hostAddress, QWidget* parent)
	:QLineEdit(parent)
{
	this->installEventFilter(this);
	this->setValidator(new LHostAddressValidator(this));
	connect(this, &LHostAddressEdit::cursorPositionChanged, this, &LHostAddressEdit::handleCursorPositionChanged);
	connect(this, &LHostAddressEdit::selectionChanged, [=]()
		{
			if (this->selectedText() == ".")
				this->setSelection(this->cursorPosition(), 0);
		});
	this->setHostAddress(hostAddress);
	//connect(this, &QLineEdit::editingFinished, this, &LHostAddressEdit::hostAddressEditFinish);
}

LHostAddressEdit::LHostAddressEdit(const QString& text, QWidget* parent)
	:LHostAddressEdit(QHostAddress(text), parent)
{
}

LHostAddressEdit::LHostAddressEdit(QWidget* parent)
	:LHostAddressEdit(QHostAddress(), parent)
{
}

void LHostAddressEdit::setHostAddress(const QHostAddress& hostAddress)
{
	if (hostAddress.isNull() || hostAddress.protocol() != QAbstractSocket::IPv4Protocol || hostAddress.toString() == _hostAddress.toString())
	{
		if (this->text() != _hostAddress.toString())
			QLineEdit::setText(_hostAddress.toString());
		return;
	}
	_hostAddress = hostAddress;
	QLineEdit::setText(_hostAddress.toString());
	emit this->hostAddressChanged(_hostAddress);
}

QHostAddress LHostAddressEdit::hostAddress() const
{
	return _hostAddress;
}

int LHostAddressEdit::currentBlock() const
{
	return _currentblock;
}

int LHostAddressEdit::defaultBlock() const
{
	return _defaultblock;
}

void LHostAddressEdit::setCurrentBlock(const int block)
{
	if (block < 0 || block > 3)
		return;
	this->setHostAddress(QHostAddress(this->text()));
	QStringList ipv4 = (this->text().split("."));
	if (ipv4[_currentblock].length() == 0)
	{
		ipv4[_currentblock] = _hostAddress.toString().split(".").at(_currentblock);
		QLineEdit::setText(ipv4.join("."));
	}
	_currentblock = block;
	ipv4 = (this->text().split("."));
	int beginPos = currentBlock();
	for (int i = 0; i < _currentblock; ++i)
	{
		beginPos += ipv4[i].length();
	}
	if (this->hasFocus())
		this->setSelection(beginPos, ipv4[_currentblock].length());
	emit this->currentBlockChanged(_currentblock);
}

void LHostAddressEdit::setDefaultBlock(const int block)
{
	if (block >= 0 && block < 4)
		_defaultblock = block;
}

void LHostAddressEdit::setText(const QString& text)
{
	QHostAddress hostAddress(text);
	this->setHostAddress(hostAddress);
}

void LHostAddressEdit::hostAddressEditFinish()
{
	this->setHostAddress(QHostAddress(this->text()));
	emit this->hostAddressEditFinished(_hostAddress);
}

void LHostAddressEdit::focusOutEvent(QFocusEvent* event)
{
	QLineEdit::focusOutEvent(event);
	hostAddressEditFinish();
}

void LHostAddressEdit::focusInEvent(QFocusEvent* event)
{
	QLineEdit::focusInEvent(event);
	QTimer::singleShot(100, [=]() {this->setCurrentBlock(_defaultblock); });
}

void LHostAddressEdit::keyPressEvent(QKeyEvent* event)
{
	if (this->hasSelectedText())
	{
		if (event->key() == Qt::Key_Left)
		{
			this->setCurrentBlock(_currentblock - 1);
			return;
		}
		else if (event->key() == Qt::Key_Right)
		{
			this->setCurrentBlock(_currentblock + 1);
			return;
		}
	}
	QLineEdit::keyPressEvent(event);
	if (event->key() == Qt::Key_Period)
	{
		this->nextBlock();
	}
}

void LHostAddressEdit::handleCursorPositionChanged(int oldPos, int newPos)
{
	if (this->hasSelectedText() && this->selectedText() == this->text())
		return;
	QStringList ipv4 = (this->text().split("."));
	if (ipv4.length() != 4) return;
	int pos = newPos;
	int newBlock;
	// 判断当前光标位置是否在第几个里面
	if (pos <= ipv4[0].length())
		newBlock = 0;
	else if (pos <= ipv4[1].length() + ipv4[0].length() + 1)
		newBlock = 1;
	else if (pos <= ipv4[2].length() + ipv4[1].length() + ipv4[0].length() + 2)
		newBlock = 2;
	else if (pos <= ipv4[3].length() + ipv4[2].length() + ipv4[1].length() + ipv4[0].length() + 3)
		newBlock = 3;
	else
		newBlock = 0;
	if (newBlock != _currentblock)
		this->setCurrentBlock(newBlock);
}

bool LHostAddressEdit::eventFilter(QObject* watched, QEvent* event)
{
	/*
	 * 这里拦截了输入法，如果输入的是.，则调用nextBlock(),因为在centos上，无法捕获到
	 * .键，所以需要在eventFilter中捕获输入的.
	 */
	if (event->type() == QEvent::InputMethod)
	{
		QInputMethodEvent* inputEvent = static_cast<QInputMethodEvent*>(event);
		// 判断是不是输入的.
		if (inputEvent->commitString() == ".")
		{
			this->nextBlock();
			return false;
		}
	}

	return QLineEdit::eventFilter(watched, event);
}

void LHostAddressEdit::nextBlock()
{
	this->setCurrentBlock(_currentblock + 1);
}

File: Python/LWidget/LLineEdit.py
from typing import overload, List
from PySide6.QtWidgets import *
from PySide6.QtCore import *
from PySide6.QtGui import *
from PySide6.QtNetwork import *


class LFileLineEdit(QLineEdit):
    fileSelected = Signal(list, str)
    fileCleared = Signal()

    class Info:
        def __init__(self, mode: QFileDialog.FileMode = QFileDialog.FileMode.ExistingFile,
                     path: str = QApplication.applicationFilePath(),
                     title: str = "请选择文件",
                     filters=None,
                     icon=None,
                     ):
            if filters is None:
                filters = ["All Files (*)"]
            if icon is None:
                self.icon = QIcon(":/res/res/file.png")
            self.mode = mode
            self.path = path
            self.filters = filters
            self.title = title

    @overload
    def __init__(self, parent: QWidget = None):
        ...

    @overload
    def __init__(self, path: str, parent: QWidget = None):
        ...

    def __init__(self, *args, **kwargs):
        if len(args) == 0:
            super().__init__()
        elif len(args) == 1:
            parent = args[0]
            super().__init__(parent)
        else:
            parent = args[1]
            text = args[0]
            super().__init__(text, parent)

        self._action: QAction = QAction()
        self._clearAction: QAction = QAction()
        self._info = LFileLineEdit.Info()
        self.setReadOnly(True)
        self._clearAction.setIcon(QIcon(":/res/res/clear.png"))
        self._clearAction.setToolTip("清空")
        self.addAction(self._clearAction, QLineEdit.ActionPosition.TrailingPosition)
        self._clearAction.triggered.connect(self.clearActionTrigger)
        self.reAction()

    def action(self) -> QAction:
        return self._action

    def clearAction(self) -> QAction:
        return self._clearAction

    def info(self) -> Info:
        return self._info

    def setInfo(self, info: Info):
        self._info = info
        self.reAction()

    @overload
    def selectedFiles(self) -> List[str]:
        ...

    @overload
    def selectedFiles(self, sep: str) -> str:
        ...

    def selectedFiles(self, *args, **kwargs):
        if len(args) == 0 and len(kwargs) == 0:
            return self.text().split(";")
        else:
            return self.text().join(args[0])

    def setClearActionVisible(self, visible: bool):
        self._clearAction.setVisible(visible)

    def clearActionVisible(self) -> bool:
        return self._clearAction.isVisible()

    @Slot()
    def showFileDialog(self):
        fileDialog = QFileDialog(None, self._info.title, self._info.path)
        fileDialog.setNameFilters(self._info.filters)
        fileDialog.setFileMode(self._info.mode)
        if fileDialog.exec() == QFileDialog.DialogCode.Accepted:
            fileList = fileDialog.selectedFiles()
            filters = fileDialog.selectedNameFilter()
            if fileList:
                self.setText(";".join(fileList))
                self.setToolTip("\n".join(fileList))
                self.fileSelected.emit(fileList, filters)

    def reAction(self):
        self._action.deleteLater()
        self._action = None
        self._action = QAction()
        self._action.setIcon(self._info.icon)
        self.setPlaceholderText(self._info.title)
        self._action.setToolTip(self._info.title)
        self.addAction(self._action, QLineEdit.ActionPosition.TrailingPosition)
        self._action.triggered.connect(self.showFileDialog)

    def clearActionTrigger(self):
        self.clear()
        self.setToolTip("")
        self.fileCleared.emit()


class LFocusSelectLineEdit(QLineEdit):
    def __init__(self, text="", parent=None):
        super().__init__(text, parent)

    def focusInEvent(self, arg__1: QFocusEvent):
        super().focusInEvent(arg__1)
        QTimer.singleShot(100, lambda: self.selectAll())


class LHostAddressValidator(QValidator):
    def __init__(self, parent=None):
        super().__init__(parent)

    def validate(self, inputText: str, pos: int) -> QValidator.State:
        regularExpression = QRegularExpression(
            "^((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])?\\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])?$")
        match: QRegularExpressionMatch = regularExpression.match(inputText)
        if match.hasMatch():
            return QValidator.State.Acceptable
        return QValidator.State.Invalid


class LHostAddressEdit(QLineEdit):
    currentBlockChanged = Signal(int)
    hostAddressChanged = Signal(QHostAddress)
    hostAddressEditFinished = Signal(QHostAddress)

    @overload
    def __init__(self, text: str = "", parent: QWidget | None = None) -> None:
        ...

    @overload
    def __init__(self, parent: QWidget | None = None) -> None:
        ...

    @overload
    def __init__(self, hostAddress: QHostAddress, parent: QWidget | None = None):
        ...

    def __init__(self, *args, **kwargs):
        self._hostAddress = QHostAddress()
        self._currentblock = 0
        self._defaultblock = 0
        parent = None
        hostAddress = QHostAddress()
        if len(args) == 1:
            parent = args[0]
        elif len(args) == 2:
            if isinstance(args[0], QHostAddress):
                hostAddress = args[0]
            else:
                hostAddress = QHostAddress(args[0])
            parent = args[1]
        super().__init__(parent)
        self.setValidator(LHostAddressValidator(self))
        self.cursorPositionChanged.connect(self.handleCursorPositionChanged)
        self.selectionChanged.connect(self.handleSelectionChanged)
        self.setHostAddress(hostAddress)

    def hostAddress(self) -> QHostAddress:
        return self._hostAddress

    def currentBlock(self) -> int:
        return self._currentblock

    def defaultBlock(self) -> int:
        return self._defaultblock

    @Slot(str)
    def setText(self, text: str):
        self.setHostAddress(QHostAddress(text))

    @Slot()
    def hostAddressEditFinish(self):
        self.setHostAddress(QHostAddress(self.text()))
        self.hostAddressEditFinished.emit(self._hostAddress)

    @Slot()
    def nextBlock(self):
        self.setCurrentBlock(self._currentblock + 1)

    @Slot(int)
    def setCurrentBlock(self, block: int):
        if 0 <= block <= 3:
            self.setHostAddress(QHostAddress(self.text()))
            ipv4: list = self.text().split(".")
            if len(ipv4[self._currentblock]) == 0:
                ipv4[self._currentblock] = len(self._hostAddress.toString().split(".")[self._currentblock])
                super().setText(".".join(ipv4))
            self._currentblock = block
            beginPos = self._currentblock + sum([len(ipv4[i]) for i in range(self._currentblock)])
            if self.hasFocus():
                self.setSelection(beginPos, len(ipv4[self._currentblock]))
            self.currentBlockChanged.emit(self._currentblock)

    @Slot(int)
    def setDefaultBlock(self, block: int):
        if 0 <= block <= 3:
            self._defaultblock = block

    @Slot(QHostAddress)
    def setHostAddress(self, hostAddress: QHostAddress):
        if hostAddress.isNull() or hostAddress.toString() == self._hostAddress.toString():
            if self.text() != self._hostAddress.toString():
                super().setText(self._hostAddress.toString())
                return
        self._hostAddress = hostAddress
        super().setText(self._hostAddress.toString())
        self.hostAddressChanged.emit(self._hostAddress)

    def focusOutEvent(self, event: QFocusEvent) -> None:
        super().focusOutEvent(event)
        self.hostAddressEditFinish()

    def focusInEvent(self, event: QFocusEvent) -> None:
        super().focusInEvent(event)
        QTimer.singleShot(100, lambda: self.setCurrentBlock(self._defaultblock))

    def keyPressEvent(self, event: QKeyEvent) -> None:
        if self.hasSelectedText():
            if event.key() == Qt.Key.Key_Left:
                self.setCurrentBlock(self._currentblock - 1)
                return
            elif event.key() == Qt.Key.Key_Right:
                self.setCurrentBlock(self._currentblock + 1)
                return
        super().keyPressEvent(event)
        if event.key() == Qt.Key.Key_Period:
            self.nextBlock()

    def handleCursorPositionChanged(self, oldPosition: int, newPosition: int) -> None:
        if self.hasSelectedText() and self.selectedText() == self.text():
            return
        ipv4 = self.text().split(".")
        pos = newPosition
        newBlock = 0

        if pos <= len(ipv4[0]):
            newBlock = 0
        elif pos <= len(ipv4[0]) + len(ipv4[1]) + 1:
            newBlock = 1
        elif pos <= len(ipv4[0]) + len(ipv4[1]) + len(ipv4[2]) + 2:
            newBlock = 2
        elif pos <= len(ipv4[0]) + len(ipv4[1]) + len(ipv4[2]) + len(ipv4[3]) + 3:
            newBlock = 3
        else:
            newBlock = 0
        if newBlock != self._currentblock:
            self.setCurrentBlock(newBlock)

    def handleSelectionChanged(self) -> None:
        if self.selectedText() == ".":
            self.setSelection(self.cursorPosition(), 0)
